feature/home/src/main/kotlin/nl/q42/template/home/main/presentation/HomeViewState.kt

```diff
@@ -2,8 +2,8 @@ package nl.q42.template.home.main.presentation
 
 import nl.q42.template.ui.presentation.ViewStateString
 
-data class HomeViewState(
-    val userEmailTitle: ViewStateString? = null,
-    val isLoading: Boolean = false,
-    val showError: Boolean = false,
-)
+sealed interface HomeViewState {
```

Comment:
I believe sealed ViewStates are inconvenient in usage. I tried it a few times, but always move away from them eventually, when not doing SDUI. Some of the reasons:

1. They presume you never load while showing content. (pull to refresh, for example, or partial loading).
2. They presume you never error (like a snackbar or any non fullscreen error) when showing old or partial content or loading.
3. It's inconvenient when you get content from different sources (multiple observables from the domain layer, callbacks from something, function calls from the view). You need to be able to create or copy the ViewState.Content (so support both: copy and create) in all locations where you want to adjust something in the ViewState's content.

Unless you have a fix for this, I believe with the old ViewState you are more flexible and have to write less code to change values in the content.

Reply:
Thanks for the feedback, I hadn't though on the perspective of SDUI vs non-SDUI.

My ideas through the reasons with your new perspective of non-SDUI apps:
1. This we solved for Hema and PostNL (SDUI), but I think the same solution applies for non-SDUI apps: the app defines clear ViewStates for Full-screen-loading vs content+loading, it would look like this:
``` kotlin
sealed interface HomeViewState {
    data class Content(val isLoadingWithPTR: Boolean, val userEmailTitle: ViewStateString) : HomeViewState
    data object FullScreenLoader : HomeViewState
    data object Error : HomeViewState
}
```
  The advantage here is that dev has to make an explicit decision about which type of loader (s)he wants to show, making the case of showing a PTR loader when there is no content impossible.

2. The same concept applies for full screen errors and transient errors on both Hema and PostNL:
``` kotlin
sealed interface HomeViewState {
    data class Content(val transientError: MyErrorType?, val userEmailTitle: ViewStateString) : HomeViewState
    data object Loader : HomeViewState
    data object FullScreenError : HomeViewState
}
```
  In this scenario, normally full-screen-errors are just a simple boolean and transient-errors (like snackbars) can contain more complex logic like a message plus an action of a more complex type.

3. For this one I wanna make sure I got point correctly, what I understand is: 
    1. for example: the ViewState has some content with 3 items that are not favorite
    2. the user favorites element 1 on the UI layer and a call to the ViewModel is triggered to change the data model
    3. the inconvenience you mention is that in the VM the dev should write something like:
``` kotlin
fun onFavoriteClicked(index: Int) {
  when(viewState) {
    is HomeViewState.Content -> {
      viewState.items[index].isFavorite = true
    }
    HomeViewState.Loading, HomeViewState.Error -> {} // no-op or throw IllegalStateException()
  }
}
```
instead of just 
``` kotlin
fun onFavoriteClicked(index: Int) {
  viewState.items?.get(index)?.isFavorite = true
}
```

Is this understanding correct? Please correct me if I misunderstood this point. Otherwise, this is my proposed solution:
Since we can only modify content once we already have content, we can start those functions with a check, either a soft one or a hard one:
``` kotlin
fun onFavoriteClicked(index: Int) {
  if (viewState !is HomeViewState.Content ) return // or throw IllegalStateException()
  viewState.items[index].isFavorite = true
}
```
This has the advantage of making the code simple and readable but still able to catch bugs faster.

Let me know what you think of these solutions and if you think we can reach a more convenient and also safe solution.